### PR TITLE
New Annotation: backend-probe-path for custom Healthprobe paths

### LINKED ISSUE #906: new annotation for custom health probe path
**Is your feature request related to a problem? Please describe.**
Currently the health probe path is obtained from the readiness probes of the pods in the deployment setup behind the ingress controller.  With Istio, the readiness probe is on port 15020 with a path of "/healthz/ready".  AGIC picks up the path from this and sends the health checks to the traffic port which then goes to the service that Istio routes traffic to.  In my case, that path is invalid and will return a 404 which will fail the health check on the application gateway.  The best solution would be if the application gateway supported health checks on a different port from the traffic port but since it doesn't, having the ability to set a custom health probe path will help make things easier.

**Describe the solution you'd like**
I would like a new annotation created with a suggested name of "appgw.ingress.kubernetes.io/backend-probe-path" where a custom probe path could be defined on the listener.

### PR DESCRIPTION
**Pull request by o-otte**
## Checklist
- [x] The title of the PR is clear and informative
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] Issues addressed by the PR are mentioned in the description followed by `Fixes`.

## Description

This PR adds a new annotion `appgw.ingress.kubernetes.io/backend-probe-path: "/path/to/healthcheck"`. It addresses the Issue #906. If the annotation is set, then the Path of the Healthprobe is overwritten. If no valid path (e.g. containing spaces), then the healthprobe path defaults to `/`.

The PR coverage contains documentation and tests.

## Fixes

#906

### CODE REVIEW

**Comment by akshaysngupta** [comment]
@o-otte Thanks for creating this PR.

Can you add a Pod YAML (along with side-car) from a cluster where Istio is installed ?

File: pkg/appgw/health_probes.go [hunk 1]
```diff
@@ -162,6 +163,16 @@ func (c *appGwConfigBuilder) generateHealthProbe(backendID backendIdentifier) *n
 		probe.Path = to.StringPtr(strings.TrimRight(*probe.Path, "*"))
 	}
 
+	if customPath, err := annotations.BackendProbePath(backendID.Ingress); err == nil {
+		if matched, reErr := regexp.MatchString(`^\/([.a-zA-Z0-9-_~\?\#\[\]\@\:\$\&\(\)\*\+\,\;\=]+/??)*$`, customPath); matched == true && reErr == nil {
+			probe.Path = to.StringPtr(customPath)
+			glog.V(5).Infof("Created custom path %s for ingress %s/%s probe", *probe.Path, backendID.Ingress.Namespace, backendID.Ingress.Name)
+		} else {
+			probe.Path = to.StringPtr("/")
+			glog.V(5).Infof("Custom path %s for ingress %s/%s probe is not valid! Using '/' instead.", customPath, backendID.Ingress.Namespace, backendID.Ingress.Name)
```
Comment (3quanfeng) [inline pkg/appgw/health_probes.go:172]: use warning instead of info

Comment (o-otte) [reply pkg/appgw/health_probes.go]: Resolved in [
d2ac719 ](https://github.com/Azure/application-gateway-kubernetes-ingress/pull/1006/commits/d2ac719c07d1ec7c3ab1701ee70a07784a86e9de)

**Review by 3quanfeng — APPROVED** [review]
LGTM!

**Comment by akshaysngupta** [comment]
@o-otte Thanks for this PR.
Here is another PR addressing the same issue adding additional useful annotations.
https://github.com/Azure/application-gateway-kubernetes-ingress/pull/1062

**Pull request closed without merging**